### PR TITLE
[receiver/k8sobjects] Support running in kubernetes leader election mode.

### PR DESCRIPTION
**Pull request by JaredTan95**
**Description:** add leader election func to support otel col running in kubernetes leader election mode

BTW, as @dmitryax https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/17369#issuecomment-1465538050, this feature also works for `k8scluster`, `k8sevents` receivers, so put this function as an `internal` method.

**Link to tracking Issue:**  https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/17369, https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32994

**Testing:** <Describe what testing was performed and which tests were added.>
1. deploy otel col in 3 replicas, and configuraed `k8sobjectsreceiver` to watch kubernetes event logs as follows yaml:
```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  namespace: default
  name: otel-event-receiver-lr
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: otel-event-receiver-lr
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: otel-event-receiver-lr
subjects:
  - kind: ServiceAccount
    name: otel-event-receiver-lr
    namespace: default
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: otel-event-receiver-lr
rules:
- apiGroups:
  - coordination.k8s.io
  resources:
  - leases
  verbs:
  - create
  - get
  - update
- apiGroups:
  - ""
  resources:
  - events
  - pods
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - "events.k8s.io"
  resources:
  - events
  verbs:
  - watch
  - list
---
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: otel-event-receiver-collector
spec:
  replicas: 3
  mode: deployment # This configuration is omittable.
  image: wutang/otelcontribcol:lr-0831 # my onw image build By PR changes.
  serviceAccount: otel-event-receiver-lr
  config: |
    receivers:
      k8sobjects:
        leader_election:
          enabled: true
          lock_name: otel-event-receiver-lr
        auth_type: serviceAccount
        objects:
          - name: events
            mode: watch
            group: events.k8s.io

    exporters:
      logging:
        verbosity: detailed

    service:
      pipelines:
        logs:
          receivers: [k8sobjects]
          exporters: [logging]
```
2.  check 3 replica pod's logs, there is always one pod in working and watch event lgos:
image

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Comment by JaredTan95** [comment]
@jpkrohling hi, do you have time to help review this PR?

File: .chloggen/support-kubernetes-leader-election.yaml
Comment (jpkrohling) [inline .chloggen/support-kubernetes-leader-election.yaml:12]:
```suggestion
note: Support running selected Kubernetes components in "leader election mode", meaning that only one replica of the Collector will react to Kubernetes events at any given point in time.
```

File: receiver/k8sobjectsreceiver/config_test.go
Comment (jpkrohling) [inline receiver/k8sobjectsreceiver/config_test.go:68]:
```suggestion
	expectedLeaderElection := k8sconfig.LeaderElectionConfig{
```

File: receiver/k8sobjectsreceiver/receiver.go
Comment (jpkrohling) [inline receiver/k8sobjectsreceiver/receiver.go:64]: Should this be prefixed with the component's name? Or is it OK to have the same leader ID for two components on the same instance?

File: receiver/k8sobjectsreceiver/factory.go
Comment (jpkrohling) [inline receiver/k8sobjectsreceiver/factory.go:35]: What's where did `e3385b9a` come from? Ideally, this should be the pod's name. Can't this be set using the in-cluster client?

**Comment by jpkrohling** [comment]
I forgot to mention during the review, but this requires a good bit of documentation, telling users what's the expectation when setting this.

File: receiver/k8sobjectsreceiver/receiver.go
Comment (JaredTan95) [reply receiver/k8sobjectsreceiver/receiver.go:64]: nice catch, I will add the component's name as electionId's prefix.

**Comment by JaredTan95** [comment]
> I forgot to mention during the review, but this requires a good bit of documentation, telling users what's the expectation when setting this.

yes, But where to hold this documentation? insert in `READEME.md`?

File: receiver/k8sobjectsreceiver/README.md
Comment (jinja2) [inline receiver/k8sobjectsreceiver/README.md:62]: nit: this should be `renew_deadline`

File: receiver/k8sobjectsreceiver/README.md
Comment (jinja2) [inline receiver/k8sobjectsreceiver/README.md:58]: I think we should mention the additional access needed by the collector's service account to perform leader election here, and also update the helm chart.

I believe this snippet should cover rbac for leaderelection.
```
- apiGroups:
  - coordination.k8s.io
  resources:
  - leases
  verbs:
  - create
  - get
  - update
  ```
  
  In case users prefer to make this stricter and are providing their own election ids, they can update the role to something like
```

- apiGroups:
  - coordination.k8s.io
  resources:
  - leases
  verbs:
  - create
- apiGroups:
  - coordination.k8s.io
  resourceNames:
  - <ELECTION ID>
  resources:
  - leases
  verbs:
  - get
  - update
  ```

Comment (JaredTan95) [reply receiver/k8sobjectsreceiver/README.md]: > and also update the helm chart.

yes, I will update soon in [opentelemetry-helm-charts](https://github.com/open-telemetry/opentelemetry-helm-charts) repo after this PR get merged. :-P

File: internal/k8sconfig/leaderelection.go
Comment (jpkrohling) [inline internal/k8sconfig/leaderelection.go:66]:
```
return "", fmt.Errorf("not running in-cluster, please specify leaderElectionIDspace")
```

What should users specify?

File: receiver/k8sobjectsreceiver/receiver.go
Comment (jpkrohling) [inline receiver/k8sobjectsreceiver/receiver.go:74]: Why this? Can't we have "component/name"? Is it a problem with slashes? If so, perhaps it's simpler to just run a strings.Replace(params.ID.String(), "/", "-") ?

Comment (jpkrohling) [reply receiver/k8sobjectsreceiver/receiver.go:74]: I would also split this into its function and write a couple of unit tests for it.

File: receiver/k8sobjectsreceiver/receiver.go
Comment (jpkrohling) [inline receiver/k8sobjectsreceiver/receiver.go:93]:
```suggestion
		kr.logger.Info("this instance of the component was selected as the current leader")
```

File: receiver/k8sobjectsreceiver/README.md
Comment (jpkrohling) [inline receiver/k8sobjectsreceiver/README.md:58]: What happens if I have two instances of this component for the same collector? Is only one of them going to be elected leader? Why would I use this? Should I always use this, or only on specific situations?

File: receiver/k8sobjectsreceiver/receiver.go
Comment (jpkrohling) [inline receiver/k8sobjectsreceiver/receiver.go:108]:
```suggestion
				kr.logger.Error("this instance of the component was previously the leader but was removed as such")
```

File: receiver/k8sobjectsreceiver/receiver.go
Comment (jpkrohling) [inline receiver/k8sobjectsreceiver/receiver.go:92]: To make it easier to test, I would split this into its own top-level function.

File: receiver/k8sobjectsreceiver/receiver.go
Comment (JaredTan95) [reply receiver/k8sobjectsreceiver/receiver.go:74]: yes, `/` is invalid in k8s reource name spec.

Comment (JaredTan95) [reply receiver/k8sobjectsreceiver/receiver.go:74]: > I would also split this into its function and write a couple of unit tests for it.

ok

File: receiver/k8sobjectsreceiver/README.md
Comment (JaredTan95) [reply receiver/k8sobjectsreceiver/README.md]: I updated the docs.

File: receiver/k8sobjectsreceiver/README.md
Comment (dmitryax) [inline receiver/k8sobjectsreceiver/README.md:61]:
```suggestion
- `leader_election`: Leader election mode helps to ensure that only one instance (as leader instance) is collecting `k8sobjects` data, and if the leader instance fails, another one is elected as leader and takes its place. 
*Note*: You need to add additional RBAC to perform the leader election mode. See below [Leader Election RBAC](#RBAC-Leader-Election) for more information.
```

File: internal/k8sconfig/leaderelection.go
Comment (JaredTan95) [reply internal/k8sconfig/leaderelection.go:38]: From my understanding , Multiple pods preempt the same resource, so there must specify the name of the resource to be preempted, and  during a pod attempts to preempt a resource lock, it carries its own unique identifier. We used the pod uid here. https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c70e0043f9145a469a4881a94fd3e5354a4316b6/internal/k8sconfig/leaderelection.go#L47
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c70e0043f9145a469a4881a94fd3e5354a4316b6/internal/k8sconfig/leaderelection.go#L57

Comment (jinja2) [reply internal/k8sconfig/leaderelection.go:38]: The lockName shouldn't be just the UID of the controller (deployment, sts, etc.) cause we can have multiple components, e.g. k8sobjectreceiver and kubernetes-cluster receiver, running in the same deployment of collector and trying to elect a leader separately. So we need the lock name to indicate which component it is for. Ideally, the lock name would be unique per component per collector installation.

The `Identity` mentioned above is the id to use to identify the lock holder. Using the pod name (hostname lookup from inside the pod will give the pod name, and not the uid), should be good enough for the id.

File: receiver/k8sobjectsreceiver/receiver.go
Comment (jinja2) [inline receiver/k8sobjectsreceiver/receiver.go:76]: So the readme above seems to imply that users can set the lock name with the option `lock_name`, but this is always overriding the user provided name and setting the lock name to the component id, right?

File: internal/k8sconfig/leaderelection.go
Comment (JaredTan95) [reply internal/k8sconfig/leaderelection.go:38]: yes, that's why I use component id as `lock name` here.

File: receiver/k8sobjectsreceiver/receiver.go
Comment (JaredTan95) [reply receiver/k8sobjectsreceiver/receiver.go:76]: Thank you for reminding me, I will update to make a supplementary logic, if the user does not set `luckname`, the default is `component id`

File: receiver/k8sobjectsreceiver/README.md
Comment (jpkrohling) [inline receiver/k8sobjectsreceiver/README.md:64]: I'm not sure what this means just by reading this doc. Are "LeaderElector clients" referred to as "non-leader candidates" in a previous item?

File: receiver/k8sobjectsreceiver/receiver.go
Comment (jpkrohling) [inline receiver/k8sobjectsreceiver/receiver.go:98]: If there are two instances of the same component, each with their own lock names, would it be clear which one is which? Perhaps add the lock name to this log entry?

**Review by jpkrohling — APPROVED** [review]
LGTM, apart from a log message that could be changed in a follow-up PR.

**Comment by JaredTan95** [comment]
> LGTM, apart from a log message that could be changed in a follow-up PR.

👌

File: internal/k8sconfig/leaderelection.go
Comment (dmitryax) [reply internal/k8sconfig/leaderelection.go:38]: @jinja2 that makes sense, but we can compose the name from UID of the controller and the component name, right?

Comment (JaredTan95) [reply internal/k8sconfig/leaderelection.go:38]: yes，BTW，could you tell me how to get the UID of collector？you meaning pod uid of collector？

Comment (JaredTan95) [reply internal/k8sconfig/leaderelection.go:38]: It is best to have an UID which multiple otel col shared used.

Comment (jinja2) [reply internal/k8sconfig/leaderelection.go:38]: Yes, we can add the uid. Easiest way to get it would be from the owner reference section of the pod, `metadata.ownerReferences[].uid`, but this will be an additional lookup. I don't feel too strongly about adding the uid to the lock name cause how likely are users to run multiple collector deployments with same component ids, and even if they do, they have the option to set the lock name to be non-colliding.

@dmitryax There is no concept of identity in the collector, right? It would be helpful to have a way for collector components to identify themself as part of a specific installation, so you can startup collectors with an id like `--class my-release` and use that for issues like this, instead of relying on the id of platform specific controllers.

Comment (JaredTan95) [reply internal/k8sconfig/leaderelection.go:38]: > It would be helpful to have a way for collector components to identify themself as part of a specific installation, so you can startup collectors with an id like `--class my-release` and use that for issues like this, instead of relying on the id of platform specific controllers.

👍 Good idea. Do I need to create an issue to track it as an discussion?

File: receiver/k8sobjectsreceiver/receiver.go
Comment (JaredTan95) [reply receiver/k8sobjectsreceiver/receiver.go:98]: luck name could not identify which the current instance is, how about pod ip or hostname ?

File: internal/k8sconfig/leaderelection.go
Comment (dmitryax) [reply internal/k8sconfig/leaderelection.go:38]: > Yes, we can add the uid. Easiest way to get it would be from the owner reference section of the pod, metadata.ownerReferences[].uid, but this will be an additional lookup

It'll be only one lookup on the start, so should not be a problem

> @dmitryax There is no concept of identity in the collector, right? It would be helpful to have a way for collector components to identify themself as part of a specific installation, so you can startup collectors with an id like --class my-release and use that for issues like this, instead of relying on the id of platform specific controllers.

There is no. Collector always rely on configuration yaml which is mostly decoupled per component. I believe this option (lock_name) should serve that purpose explicitly, but it'd be nice not to make it required for users.

**Comment by jpkrohling** [comment]
@dmitryax, is there anything else you'd want from this PR?

File: internal/k8sconfig/leaderelection.go
Comment (dmitryax) [reply internal/k8sconfig/leaderelection.go:38]: @JaredTan95 can you please create an issue to follow up on this?

File: receiver/k8sobjectsreceiver/receiver.go
Comment (dmitryax) [inline receiver/k8sobjectsreceiver/receiver.go:78]: I don't think this is a valid fallback. This way, the leader election will be shared across different collector deployments, deamonsets or statefulsets. If you don't want to implement https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24662/files#r1291903892 for now, I think we need to make the the lock name field required for the end user

File: receiver/k8sobjectsreceiver/receiver.go
Comment (dmitryax) [inline receiver/k8sobjectsreceiver/receiver.go:117]: We should never initialize `leaderElectionClient` if user doesn't enable it, so I'd remove the first condition to avoid confusion

File: receiver/k8sobjectsreceiver/receiver.go
Comment (JaredTan95) [reply receiver/k8sobjectsreceiver/receiver.go:78]: yes, `UID` of collector could be implemented by another separate followed-up PR

File: receiver/k8sobjectsreceiver/receiver.go
Comment (newly12) [inline receiver/k8sobjectsreceiver/receiver.go:106]: I think the receiver should stop collecting object when it is not the leader any more, and go back into the _wait_ mode?

Comment (JaredTan95) [reply receiver/k8sobjectsreceiver/receiver.go:106]: You are right, but once a pod is running as a leader, it will always be the leader, unless it crashes, but when it crashes, other replicas will attempt to become the leader, so the new pod(crashed before) will become a backup node, naturally in wait mode.

Comment (jinja2) [reply receiver/k8sobjectsreceiver/receiver.go:106]: > once a pod is running as a leader, it will always be the leader, unless it crashes
I don't think we should rely on this assumption. The non-leader candidates will wait the `LeaseDuration` (default 15sec) after the last successful renewal by leader to attempt to acquire the lock. It is possible that the leader pod is unable to renew the lease object for a short duration of time due to issues like network partition/congestion, etc. Once the old leader is back (as in can connect to k8s api again), another candidate could very well have acquired leadership

Comment (JaredTan95) [reply receiver/k8sobjectsreceiver/receiver.go:106]: thx @newly12  @jinja2 , I will update the logical here

File: internal/k8sconfig/leaderelection.go
Comment (JaredTan95) [reply internal/k8sconfig/leaderelection.go:38]: > @JaredTan95 can you please create an issue to follow up on this?

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/26367

File: internal/k8sconfig/leaderelection.go
Comment (dmitryax) [inline internal/k8sconfig/leaderelection.go:37]: Please move this check to `func (*LeaderElectionConfig) Validate()` which also should have checks for invalid/negative duration fields

File: internal/k8sconfig/leaderelection.go
Comment (dmitryax) [inline internal/k8sconfig/leaderelection.go:107]: I don't think the conditions are needed. You can just assign the to `LeaderElectionConfig` fields given that it's initialize with zero values anyway

File: receiver/k8sobjectsreceiver/receiver.go
Comment (dmitryax) [inline receiver/k8sobjectsreceiver/receiver.go:98]: We need to add id (type/name) of the component in all the messages

File: receiver/k8sobjectsreceiver/receiver.go
Comment (dmitryax) [inline receiver/k8sobjectsreceiver/receiver.go:105]: Unclear how does it work given that the channel is closed in the Shutdown method. Can you please add a test covering start/shutdown callbacks?

Comment (JaredTan95) [reply receiver/k8sobjectsreceiver/receiver.go]: It's complicated to test the start/stop function of the leader elector, so I'll go ahead and explore it

File: receiver/k8sobjectsreceiver/receiver.go
Comment (JaredTan95) [reply receiver/k8sobjectsreceiver/receiver.go]: okay

**Comment by JaredTan95** [comment]
> What's the state of this PR?

finding a way to add Unit-test...

**Pull request closed without merging**
